charms/kubernetes_snaps.py

```diff
@@ -2,12 +2,393 @@
+):
+    controller_opts = {}
+
+    controller_opts["min-resync-period"] = "3m"
+    controller_opts["v"] = "2"
+    controller_opts["root-ca-file"] = "/root/cdk/ca.crt"
+    controller_opts["kubeconfig"] = kubeconfig
+    controller_opts["authorization-kubeconfig"] = kubeconfig
+    controller_opts["authentication-kubeconfig"] = kubeconfig
+    controller_opts["use-service-account-credentials"] = "true"
+    controller_opts["service-account-private-key-file"] = "/root/cdk/serviceaccount.key"
+    controller_opts["tls-cert-file"] = "/root/cdk/server.crt"
+    controller_opts["tls-private-key-file"] = "/root/cdk/server.key"
+    controller_opts["cluster-name"] = cluster_name
+    controller_opts["terminated-pod-gc-threshold"] = "12500"
+    controller_opts["profiling"] = "false"
+    controller_opts["service-cluster-ip-range"] = service_cidr
+    if cluster_cidr:
+        controller_opts["cluster-cidr"] = cluster_cidr
+    feature_gates = ["RotateKubeletServerCertificate=true"]
+
+    # TODO: cloud config
+    """
+    cm_cloud_config_path = cloud_config_path("kube-controller-manager")
+    if has_external_cloud_provider():
+        controller_opts["cloud-provider"] = "external"
+    elif is_state("endpoint.aws.ready"):
+        if kube_version < (1, 27, 0):
+            controller_opts["cloud-provider"] = "aws"
+        else:
+            hookenv.log(
+                "AWS cloud-provider is no longer available in-tree. "
+                "the out-of-tree provider is necessary",
+                level="WARNING",
+            )
+        if kube_version < (1, 25, 0):
+            feature_gates.append("CSIMigrationAWS=false")
+    elif is_state("endpoint.gcp.ready"):
+        controller_opts["cloud-provider"] = "gce"
+        controller_opts["cloud-config"] = str(cm_cloud_config_path)
+        if kube_version < (1, 25, 0):
+            feature_gates.append("CSIMigrationGCE=false")
+    elif is_state("endpoint.vsphere.ready"):
+        if (1, 12) <= kube_version:
+            controller_opts["cloud-provider"] = "vsphere"
+            controller_opts["cloud-config"] = str(cm_cloud_config_path)
+        if kube_version < (1, 26, 0):
+            feature_gates.append("CSIMigrationvSphere=false")
+    elif is_state("endpoint.azure.ready"):
+        controller_opts["cloud-provider"] = "azure"
+        controller_opts["cloud-config"] = str(cm_cloud_config_path)
+        if kube_version < (1, 25, 0):
+            feature_gates.append("CSIMigrationAzureDisk=false")
+    """
+
+    controller_opts["feature-gates"] = ",".join(feature_gates)
+
+    configure_kubernetes_service(
+        "kube-controller-manager",
+        controller_opts,
+        extra_args_config,
+    )
+
+
+def configure_kubernetes_service(service, base_args, extra_args_config):
+    extra_args = parse_extra_args(extra_args_config)
+
+    args = {}
+    args.update(base_args)
+    args.update(extra_args)
+
+    # TODO: CIS arg handling???
```

Comment:
we'll likely drop this going forward as we're cis compliant ootb and moving to trivy for runtime scanning.